File: 066.LeetCode@86_Partition_List/066.LeetCode@86_Partition_List.py
class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        less_dummy = ListNode(0)
        less_tail = less_dummy

        greater_dummy = ListNode(0)
        greater_tail = greater_dummy

        current = head

        while current:
            if current.val < x:
                less_tail.next = current
                less_tail = current
            else:
                greater_tail.next = current
                greater_tail = current
            current = current.next

        greater_tail.next = None

        less_tail.next = greater_dummy.next

        return less_dummy.next

File: 066.LeetCode@86_Partition_List/066.LeetCode@86_Partition_List.java
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode lessDummy = new ListNode(0);
        ListNode lessTail = lessDummy;

        ListNode greaterDummy = new ListNode(0);
        ListNode greaterTail = greaterDummy;

        ListNode current = head;

        while (current != null) {
            if (current.val < x) {
                lessTail.next = current;
                lessTail = current;
            } else {
                greaterTail.next = current;
                greaterTail = current;
            }
            current = current.next;
        }

        greaterTail.next = null;
        lessTail.next = greaterDummy.next;
        return lessDummy.next;
    }
}